File: src/shmif/platform/eventqueue.c
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>
#include <limits.h>
#include <stdio.h>

#include "arcan_shmif.h"
#include "shmif_platform.h"
#include "arcan_shmif_event.h"

void shmif_platform_setevqs(
	struct arcan_shmif_page* dst,
	sem_t* esem,
	struct arcan_evctx* inq, struct arcan_evctx* outq
)
{
	inq->synch.synch = &dst->esync;
	inq->synch.killswitch = &dst->dms;
	outq->synch.killswitch = &dst->dms;

#ifdef ARCAN_SHMIF_THREADSAFE_QUEUE
	if (!inq->synch.init){
		inq->synch.init = true;
		pthread_mutex_init(&inq->synch.lock, NULL);
	}
	if (!outq->synch.init){
		outq->synch.init = true;
		pthread_mutex_init(&outq->synch.lock, NULL);
	}
#endif

	inq->local = false;
	inq->eventbuf = dst->childevq.evqueue;
	inq->front = &dst->childevq.front;
	inq->back  = &dst->childevq.back;
	inq->eventbuf_sz = PP_QUEUE_SZ;

	outq->local =false;
	outq->eventbuf = dst->parentevq.evqueue;
	outq->front = &dst->parentevq.front;
	outq->back  = &dst->parentevq.back;
	outq->eventbuf_sz = PP_QUEUE_SZ;
}

File: src/shmif/platform/eventqueue.zig
// Zig reimplementation of platform/eventqueue.c
// Drop-in C-ABI-compatible replacement for event queue setup.
//
// Exports: shmif_platform_setevqs
//
const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);
const off = @import("shmif_offsets");
const c = @import("shmif_types");

export fn shmif_platform_setevqs(
    dst: *anyopaque,
    esem: ?*anyopaque,
    inq: *c.struct_arcan_evctx,
    outq: *c.struct_arcan_evctx,
) void {
    if (is_freestanding) return;
    _ = esem;

    inq.*.synch.synch = @ptrCast(@alignCast(off.Page.getEsyncPtr(dst)));
    inq.*.synch.killswitch = off.Page.getDmsPtr(dst);
    outq.*.synch.killswitch = off.Page.getDmsPtr(dst);

    // ARCAN_SHMIF_THREADSAFE_QUEUE path omitted (typically not compiled)

    inq.*.local = 0; // false
    inq.*.eventbuf = @ptrCast(@alignCast(off.Page.childevqEventbuf(dst)));
    inq.*.front = off.Page.childevqFrontPtr(dst);
    inq.*.back = off.Page.childevqBackPtr(dst);
    inq.*.eventbuf_sz = c.PP_QUEUE_SZ;

    outq.*.local = 0; // false
    outq.*.eventbuf = @ptrCast(@alignCast(off.Page.parentevqEventbuf(dst)));
    outq.*.front = off.Page.parentevqFrontPtr(dst);
    outq.*.back = off.Page.parentevqBackPtr(dst);
    outq.*.eventbuf_sz = c.PP_QUEUE_SZ;
}
